File: scripts/validate_yaml.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

const repoRoot = path.resolve(__dirname, '..');
const workflowPath = path.join(repoRoot, '.github', 'workflows', 'docker-build-push.yml');
const reportPath = path.join(repoRoot, 'docs', 'YAML_WORKFLOW_VALIDATION.md');

let content = null;
let errors = [];
try {
  content = fs.readFileSync(workflowPath, 'utf8');
} catch (err) {
  errors.push(`Failed to read workflow file: ${err.message}`);
}

if (content !== null) {
  try {
    // try parsing all documents
    const docs = yaml.loadAll(content);
    // optionally, we could do additional structure checks here
  } catch (err) {
    errors.push(err.toString());
  }
}

const now = new Date().toISOString();
const lines = [];
lines.push('# YAML Workflow Validation Report');
lines.push('');
lines.push(`- File: \\`${workflowPath}\\``);
lines.push(`- Checked at (UTC): ${now}`);
lines.push('');
if (errors.length === 0) {
  lines.push('## Result: No parse errors found');
  lines.push('');
} else {
  lines.push('## Result: Errors found');
  lines.push('');
  lines.push('### Errors');
  lines.push('');
  errors.forEach((e, i) => {
    lines.push(`${i+1}. \`\`\``);
    lines.push(e);
    lines.push('```');
    lines.push('');
  });
}

lines.push('## Workflow content (first 200 lines)');
lines.push('');
if (content !== null) {
  const workflowLines = content.split(/\r?\n/);
  const maxLines = 200;
  for (let i = 0; i < Math.min(workflowLines.length, maxLines); i++) {
    // escape backticks
    lines.push('    ' + workflowLines[i].replace(/`/g, "'") );
  }
  if (workflowLines.length > maxLines) lines.push('    ... (truncated)');
} else {
  lines.push('Unable to read workflow file.');
}

// ensure docs dir exists
try {
  fs.mkdirSync(path.dirname(reportPath), { recursive: true });
  fs.writeFileSync(reportPath, lines.join('\n'), 'utf8');
  console.log('Validation complete. Report written to', reportPath);
  if (errors.length > 0) process.exitCode = 2;
} catch (err) {
  console.error('Failed to write report:', err);
  process.exitCode = 3;
}

File: scripts/validate_yaml.py
#!/usr/bin/env python3
import sys
import os
from datetime import datetime

repo_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
workflow_path = os.path.join(repo_root, '.github', 'workflows', 'docker-build-push.yml')
report_path = os.path.join(repo_root, 'docs', 'YAML_WORKFLOW_VALIDATION.md')

try:
    import yaml
except Exception:
    # try to install pyyaml locally
    import subprocess
    print('PyYAML not found, attempting to install via pip...')
    subprocess.check_call([sys.executable, '-m', 'pip', 'install', '--user', 'PyYAML'])
    import yaml

issues = []
content = None
try:
    with open(workflow_path, 'r', encoding='utf-8') as f:
        content = f.read()
    # safe_load_all to detect all documents
    docs = list(yaml.safe_load_all(content))
    # No parse errors
except Exception as e:
    issues.append(str(e))

# prepare report
now = datetime.utcnow().isoformat() + 'Z'
lines = []
lines.append('# YAML Workflow Validation Report')
lines.append('')
lines.append(f'- File: `{workflow_path}`')
lines.append(f'- Checked at (UTC): {now}')
lines.append('')
if not issues:
    lines.append('## Result: No parse errors found')
    lines.append('')
else:
    lines.append('## Result: Errors found')
    lines.append('')
    lines.append('### Errors')
    lines.append('')
    for i, it in enumerate(issues, 1):
        lines.append(f'{i}. ```')
        lines.append(it)
        lines.append('```')
        lines.append('')

# include the workflow content for reference (truncated if long)
lines.append('## Workflow content (first 200 lines)')
lines.append('')
if content is not None:
    workflow_lines = content.splitlines()
    max_lines = 200
    for ln in workflow_lines[:max_lines]:
        # escape backticks
        lines.append('    ' + ln.replace('`', "'"))
    if len(workflow_lines) > max_lines:
        lines.append('    ... (truncated)')
else:
    lines.append('Unable to read workflow file.')

# write report
os.makedirs(os.path.dirname(report_path), exist_ok=True)
with open(report_path, 'w', encoding='utf-8') as f:
    f.write('\n'.join(lines))

print('Validation complete. Report written to', report_path)
if issues:
    sys.exit(2)
else:
    sys.exit(0)
